### ISSUE TITLE
libzstd-related failure in `main`

### ISSUE THREAD

**Issue by WardF** [Unidata/netcdf-fortran#369]
## Description 

Roadblock for `v4.6.0` release.

I've encountered a failure that appears to have been introduced in a9c83516ec50f738c93dccaa4354c18c5ae19309.  Any insight, @edwardhartnett, would be appreciated :). 

The issue is occurring on 64-bit linux (Debian & Ubuntu, so far).  It occurs when `libnetcdf` was built with `Zstd` support.  It occurs when building with `autoconf`-based tools.

It does not occur when libnetcdf was built without `Zstd` support, nor does it occur when using cmake.  It also does not occur on MacOS.  

## Symptom 

The following tests fail when running `make check`

* `ftst_vars7`
* `f03tst_vars7`

### `ftst_vars7` output:

````
wfisher@debian64:~/Desktop/gitprojects/netcdf-fortran/nf_test4$ ./ftst_vars7
 
 *** Testing quantize feature.
 *** (and zstandard compression)
 NetCDF: Filter error: undefined filter encountered                              
STOP 3
````

### `f03tst_vars7` output:

````
wfisher@debian64:~/Desktop/gitprojects/netcdf-fortran/nf_test4$ ./f03tst_vars7 
 
 *** Testing quantize feature.
 *** (and zstandard compression)
 NetCDF: Filter error: undefined filter encountered                              
STOP 3
```

**Comment by WardF** [comment]
* [config.log.txt](https://github.com/Unidata/netcdf-fortran/files/9116508/config.log.txt)

**Comment by edwardhartnett** [comment]
Is HDF5_FILTER_,PATH set?

**Comment by edwardhartnett** [comment]
When HDF5_PLUGIN_PATH is not set to the directory that contains the zstandard plugin, then the netcdf-fortran zstd test will fail.

In other words, in the CI, when setting the location of the netcdf-c library, also set HDF5_PLUGIN_PATH

**Comment by DennisHeimbigner** [comment]
The underlying problem is that in netcdf-c, these tests are wrapped inside
a shell script and that script sets HDF5_PLUGIN_PATH. Not sure how to
do equivalent in Fortran testing.b

**Comment by edwardhartnett** [comment]
It's not the same for netcdf-fortran. And we want this to fail.

The netcdf-fortran test is just the first Fortran program that they will run which will use zstandard. (And we do check that netcdf-c was built with zstandard before running the Fortran test.)

So the user wants zstandard. However, they have not yet figured out that they need to set HDF5_PLUGIN_PATH. Remember, the vast majority of installers will not read any documentation before installing. ;-)

So perhaps we should change the name of the test, and the error message it gives, to explain this better. I will submit a PR...

**Comment by edwardhartnett** [comment]
OK, I've changed it so that when it fails it says:

 ```
*** Testing quantize feature.
 *** (and zstandard compression)
 Zstandard filter not found.
 Set HDF5_PLUGIN_PATH and try again.
STOP 5

```

**Comment by WardF** [comment]
The expanded error messages are great, thanks! I'll get that merged ASAP.  I'm looking at a way to catch this during configuration, instead of expecting people to go through a cycle of 'configure->make check->re-configure-make check'.  It would make more sense, I think, for `configure` or `cmake` to highlight the need to specify this environmental variable.  Similarly, `nc-config` should provide this information as well via `nc-config --plugindir`.  That will make automating this much easier.  Perhaps this change is already in `main`, I'll double check.  

I'll get this merged and the v4.6.0 release out ASAP.  Thanks!

**Comment by edwardhartnett** [comment]
That is a good idea about nc-config, not sure if it is implemented...

**Comment by WardF** [comment]
I don't think it is, but if I go check I'll get distracted and I'd rather knock out these checks.  I have a complimentary branch I'll put up a PR against, that incorporates your great work, and also a check in `configure` and (as soon as I verify it does what I want it to do), `cmake`.

**Comment by DennisHeimbigner** [comment]
Ed  I don't understand. The test is run before install so the HDF5 filters
will not have been placed in the default HDF5_PLUGIN_PATH location yet.

**Comment by edwardhartnett** [comment]
The test is run before netcdf-fortran install, but that is after netcdf-c install. So netcdf-c has installed the filters (on the users systems, and also in the CI systems), before netcdf-fortran is built and tested.

**Comment by bzfbmsch** [comment]
I am seeing the same issue with 4.6.0. My hdf5 (1.12.1) does not have any plugins. The libzstd comes in, since the configure script for netcdf-c checks, if this library is installed in the system and it is there together with the corresponding .h-files.

If the netcdf-c library does not call methods implemented in libzstd directly, it should not link this library. Its function should come in via the hdf5-library if needed. Lets check for unneeded libraries.
ldd -u libnetcdf.so gives (among many others)
        /sw/data/hdf5/OS_15.4/x86_64-suse-linux/1.12.1/intel_2022/lib64/libhdf5_hl.so.200
        /lib64/libpthread.so.0
        /lib64/libdl.so.2
        /lib64/libm.so.6
        /lib64/libz.so.1
        /usr/lib64/libzip.so.5
        /sw/data/hdf5/OS_15.4/x86_64-suse-linux/1.12.1/intel_2022/lib64/libsz.so.2
        /sw/data/hdf5/OS_15.4/x86_64-suse-linux/1.12.1/intel_2022/lib64/libaec.so.0
        /usr/lib64/libbz2.so.1
        **/usr/lib64/libzstd.so.1**
        /usr/lib64/libblosc.so.1
        /usr/lib64/libcurl.so.4

This means some degree of overlinking, those libs are never (if ldd is right) called in libnetcdf itself but are called possibly in libhdf5. No need to link them here. This has also consequnces for nc-config and so on.

I am aware, this comment tends to overload this thread. But the request for the plugin path, if there are no plugins at all is related to this issue.

**Comment by edwardhartnett** [comment]
Can you try with netcdf-c 4.9.0? It should handle zstd well...

**Comment by bzfbmsch** [comment]
I am using netcdf-c 4.9.1-rc1, since one test was hanging in netcdf-c 4.9.0 [https://github.com/Unidata/netcdf-c/issues/2535]. This is resolved with 4.9.1 now. 
Setting a dummy path HDF5_PLUGIN_PATH lets the test core dump. This could be expected, since the plugin does not exist. 
--disable-zstandard-plugin helps to suppress usage of libzstd in the netcdf-fortran library. All tests pass through now. But this is more a workaround than a solution. It would be nice, if someone could find time to consider my previous comment on overlinking. If wanted, I will find more time for a detailed discussion.
Cheers, Martin

**Comment by DennisHeimbigner** [comment]
Of course, the reason we check for various compressor libraries
like libzstd is for use with NCZarr. But it is a fair point that 
If NCZarr is disabled and HDF5 is enabled, then we should
suppress any plugins that we can find that HDF5 is already
supporting. But as far as I know, the only two compressors
built into HDF5 are zlib an szip.
